ccf_transaction_processor/Readme.md

```diff
@@ -3,40 +3,175 @@ Licensed under Creative Commons Attribution 4.0 International License
+cmake -GNinja -DCOMPILE_TARGETS=virtual -DBUILD_END_TO_END_TESTS=OFF -DBUILD_SMALLBANK=OFF \
+                -DBUILD_TESTS=OFF-DBUILD_UNIT_TESTS=OFF ..
+```
+
+D. Build using ninja
+
+```bash
+ninja
+```
+
+If the VM has a small memory (< 4GB), it might be required to reduce the parallelism in the build process using the `-j` flag for the `ninja` command, say ninja -j2 or even ninja -j1 (the default seems to be ninja -j4).
+
+If build is successful, the libpdoenc.virtual.so library is created inside the build folder (ignore any other targets that get created)
+
+## Deploy CCF with PDO-TP as the application
+
+The following commands start a single node CCF under the virtual enclave mode.  PDO-TP is the hosted application.
+
+```bash
+cd $PDO_SOURCE_ROOT/ccf_transaction_processor/CCF/build
+python3.7 -m venv env
+source env/bin/activate
+pip install -q -U -r ../tests/requirements.txt
+python ../tests/start_network.py --gov-script ../src/runtime_config/gov.lua  --label pdo_tp -e virtual --package libpdoenc.virtual.so --node <ip-address:6006>
+```
+
+A single node CCF instance hosting the PDO TP will be available for business use @ ip-address:6006. Here `ip-address` is the address of this VM that can be used to ping this machine from other VMs (this is our use case). Set ip-address to 127.0.0.1 for local testing. If you are behind a proxy, add ip-address to the list of no_proxy bash env variable.
+
+## Share CCF (TLS) Authentication Keys
+
+CCF uses mutually authenticated TLS channels for transactions. Keys are located at
+$PDO_SOURCE_ROOT/ccf_transaction_processor/CCF/build/workspace/pdo_tp_common. The network certificate is `networkcert.pem`. User public certificate is `user0_cert.pem` and private key is `user0_privk.pem`.
+Note that the keys are created as part of CCF deployment and are unique to the specific instance of CCF.
+
+In our usage, CCF users are PDO clients, and PDO client authentication is implemented within the transaction processor itself. Thus, we do not utilize the client authentication feature provided by CCF. However to satisfy the CCF's requirement that only authorized CCF users can submit transactions to CCF, share `user0_cert.pem` and `user0_privk.pem` with all the PDO clients. These two keys and the network certificate `networkcert.pem` must be stored under $PDO_LEDGER_KEY_ROOT (as part of PDO deployment). The user keys must be renamed as `userccf_cert.pem` and `userccf_privk.pem` respectively.
+
+(The deployment script `start_network.py` generates to two additional pairs of user keys. We simply ignore these in our usage.)
+
+## Test the Deployment with Ping Test
+
+PDO TP contains a simple `ping` rpc that returns success every time it is queried. Test the PDO-TP deployment using this rpc. Invoke the following commands to issue 100 ping rpcs. The net througput is reported at the end of the test.
+
+```bash
+cd $PDO_SOURCE_ROOT/ccf_transaction_processor/test
+./test.sh <ip-address>
```

Comment:
when i tried this i got a python exception, about missing module infra. i have activated the virtual environment.

Reply:
may i ask if you did `cd`  into the directory before executing the test.sh. Infra is part of ccf's test module. There is hardcoding of paths, I can't think of any other reason why infra is not found. venv activation is taken care of by the test.sh script as well.